File: src/dist.f
C     closestdistXY,  distance between x and x or between x and y
C
C
C     We have four distances implemented:
C       c("euclidean", "maximum", "minkowski", "greatcircle")
C
c     In case we need the distance matrix between x and x, then the
c     following parameters are used as well:
c     if part=-1, lower tri, part=0 the entire matrix
c        part= 1, upper tri only.
c     only values smaller than eta are considered.
c     p power for minkowski



      double precision function euclid(x,y,p)
      implicit none
      double precision x,y,p
      p=p ! to avoid: Warning: Unused dummy argument 'p' at (1)
      euclid=(x-y)**2
      return
      end



      double precision function minkowski(x,y,p)
      implicit none
      double precision x,y,p
      minkowski=abs(x-y)**p
      return
      end



      subroutine closestdist( ncol, x,nrowx, y, nrowy,
     &    part, p, method,
     &    eta, colindices, rowpointers, entries, nnz, iflag)

      implicit none


      double precision euclid, minkowski
      external euclid, minkowski

      integer ncol,nrowx, nrowy, nnz, method,  part,  iflag
      integer colindices(nnz), rowpointers(nrowx+1)
      double precision p, x(nrowx,ncol),y(nrowy,ncol)
      double precision eta, entries(nnz)




      if (method.eq.1) then
         p=2.0
         call closestEdistXY( ncol, x,nrowx, y, nrowy,
     &        part, p, euclid,
     &        eta, colindices, rowpointers, entries, nnz, iflag)
      endif
      if (method.eq.2) then
         p=1.0
         call closestMAXdistXY( ncol, x,nrowx, y, nrowy,
     &        part,
     &        eta, colindices, rowpointers, entries, nnz, iflag)
      endif
      if (method.eq.3) then
         call closestEdistXY( ncol, x,nrowx, y, nrowy,
     &        part, p, minkowski,
     &        eta, colindices, rowpointers, entries, nnz, iflag)
      endif
      if (method.eq.4) then
         call closestGCdistXY( x,nrowx, y, nrowy,
     &        part, p,
     &        eta, colindices, rowpointers, entries, nnz, iflag)
      endif
      return
      end



      subroutine closestEdistXY( ncol, x,xnrow, y, ynrow,
     &    part, p, distfcn,
     &    eta, colindices, rowpointers, entries, nnz, iflag)

      implicit none

      double precision distfcn
      external distfcn

      integer ncol,xnrow, ynrow, nnz,  part,iflag
      integer colindices(nnz),rowpointers(xnrow+1)
      double precision p,x(xnrow,ncol), y(ynrow,ncol)
      double precision eta, entries(nnz)

c     local variables
      integer jja, i,j,k, jfrom, jto
      double precision etap, tmp,pinv


      etap=eta**p
      pinv=1/p

      jja=1

      rowpointers(1)=1
      jfrom = 1
      jto = ynrow


c cycle over all rows of x (independent of part)

      do i= 1,xnrow

         if (part .lt. 0) then
            jto = i
         endif
         if (part .gt. 0) then
            jfrom = i
         endif

         do 10 j = jfrom,jto

c Start calculating the distance (until delta is exceeded)
            tmp = 0.0
            do  k = 1, ncol
               tmp = tmp + distfcn(x(i,k),y(j,k),p)
               if( tmp.gt.etap) goto 10
            enddo
c Delta is not exceeded.

c     in case nnz was too small, recall line to get a better estimate
            if( jja .gt. nnz) then
               iflag = i
               goto 20
            endif

            colindices(jja) = j
            if (abs(p-2) .le. 0.D0) then
               entries(jja) = sqrt(tmp)
            else
               if (abs(p-1) .le. 0.D0) then
                  entries(jja) = tmp
               else
                  entries(jja) = tmp**pinv
               endif
            endif
            jja = jja + 1


 10      continue
         rowpointers(i+1)=jja
      enddo


      if (part.gt.0) then
         rowpointers(xnrow+1)=jja
      endif
      nnz=jja-1
 20   continue

      return
      end

      subroutine closestMAXdistXY( ncol, x,xnrow, y, ynrow,
     &    part,
     &    eta, colindices, rowpointers, entries, nnz, iflag)

      implicit none

      integer ncol,xnrow, ynrow, nnz,  part,iflag
      integer colindices(nnz),rowpointers(xnrow+1)
      double precision x(xnrow,ncol), y(ynrow,ncol)
      double precision eta, entries(nnz)

c     local variables
      integer jja, i,j,k, jfrom, jto
      double precision  tmp



      jja=1

      rowpointers(1)=1
      jfrom = 1
      jto = ynrow


      do i= 1,xnrow

         if (part .lt. 0) then
            jto = i
         endif
         if (part .gt. 0) then
            jfrom = i
         endif

         do 10 j = jfrom,jto

c Start calculating the distance
            tmp = 0.0
            do  k = 1, ncol
               tmp = max(tmp, abs(x(i,k)-y(j,k)))
               if( tmp.gt.eta) goto 10
            enddo

c Delta is not exceeded.
c     (i,j) has a distance smaller than eta.



c     in case nnz was too small, recall line to get a better estimate
            if( jja .gt. nnz) then
               iflag = i
               goto 20
            endif

            colindices(jja) = j
            entries(jja) = tmp
            jja = jja + 1


 10      continue
         rowpointers(i+1)=jja
      enddo


      if (part.gt.0) then
         rowpointers(xnrow+1)=jja
      endif
      nnz=jja-1
 20   continue

      return
      end



      subroutine closestGCdistXY( x,nx, y, ny,
     &    part,p,
     &    eta, colindices, rowpointers, entries, nnz, iflag)

      implicit none

      integer nx, ny, nnz, colindices(nnz),rowpointers(nx+1)
      integer part, iflag
      double precision x(nx,2), y(ny,2), p, eta, entries(nnz)

c     local variables
      logical equi
      integer jja, i,j, jfrom, jto
      double precision etap, tmp, tmp1, tmp2
      double precision rad, thres
      double precision scy12(ny), ccy12(ny), sy2(ny)
      double precision scx12,     ccx12,     sx2

      parameter (rad = 0.01745329251994329)
      parameter (thres = 0.99999999999)


c     Great savings if we know that x=y.
c     Changes for archaic d... to ... trigonometric fcn
      if (p .lt. 0) then
         equi=.TRUE.
         p=-p
      else
         equi= .FALSE.
      endif

      jja=1

      etap=cos(eta*rad)
      rowpointers(1)=1
      jfrom = 1
      jto = ny


      DO j=1,ny
         tmp1=y(j,1)*rad
         tmp2=y(j,2)*rad
         ccy12(j)=cos(tmp1)*cos(tmp2)
         scy12(j)=sin(tmp1)*cos(tmp2)
         sy2(j)=sin(tmp2)
      ENDDO



      do i= 1,nx

c     x2 is missing if equi=.TRUE. and we reuse the y stuff
         if (equi .eqv. .TRUE.) then
            ccx12=ccy12(i)
            scx12=scy12(i)
            sx2=sy2(i)
         else
            tmp1=x(i,1)*rad
            tmp2=x(i,2)*rad
            ccx12=cos(tmp1)*cos(tmp2)
            scx12=sin(tmp1)*cos(tmp2)
            sx2=sin(tmp2)
         endif

         if (part .lt. 0) then
            jto = i
         endif
         if (part .gt. 0) then
            jfrom = i
         endif

         do 10 j = jfrom,jto


c     Start calculating the distance
            tmp = ccx12 * ccy12(j) + scx12 * scy12(j)  + sx2*sy2(j)


            if (tmp .lt. etap) goto 10
c     Delta is not exceeded.

c     Due to numerical instabilities, we need the following... 0.15-2:
c     Patch suggested at code clinics.
            if  (tmp .ge. thres) then
               tmp = 0.0
            else
               tmp = acos( tmp)
            endif

c     (i,j) has a distance smaller than eta.

c     In case nnz was too small, recall line to get a better estimate
            if( jja .gt. nnz) then
               iflag = i
               goto 20
            endif

            colindices(jja) = j
            entries(jja) = tmp*p
            jja = jja + 1


 10      continue
         rowpointers(i+1)=jja
      enddo


      if (part.gt.0) then
         rowpointers(nx+1)=jja
      endif
      nnz=jja-1
 20   continue

      return
      end

File: R/dist.R
# HEADER ####################################################
# This is file spam/R/dist.R.                               #
# It is part of the R package spam,                         #
#  --> https://CRAN.R-project.org/package=spam              #
#  --> https://CRAN.R-project.org/package=spam64            #
#  --> https://git.math.uzh.ch/reinhard.furrer/spam         #
# by Reinhard Furrer [aut, cre], Florian Gerber [aut],      #
#    Roman Flury [aut], Daniel Gerber [ctb],                #
#    Kaspar Moesinger [ctb]                                 #
# HEADER END ################################################


### in base:
# dist(x, method = "euclidean", diag = FALSE, upper = FALSE, p=2)
#
### in fields
# rdist( x1, x2)
# rdist.earth(x1, x2, miles = TRUE, R = NULL)
# fields.rdist.near( x1, x2, delta, max.points= NULL)
#
### in sp 
# spDistsN1(pts, pt, longlat=FALSE)
#
### in amap     ### nbproc  integer, Number of subprocess for parallelization
# Dist(x, method = "euclidean", nbproc = 1, diag = FALSE, upper = FALSE)
#
### in argosfilter
# distance(lat1, lat2, lon1, lon2)  # gc between two pts in km
# distanceTrack(lat,lon)            # gc between pts  in km
#
### in proxy
# dist(x, y = NULL, method = NULL, ..., diag = FALSE, upper = FALSE,
#     pairwise = FALSE, by_rows = TRUE, convert_similarities = TRUE,
#     auto_convert_data_frames = TRUE)
#
### in RFOC
#  GreatDist(LON1, LAT1, LON2, LAT2, EARTHRAD= 6371)
#

nearest.dist <- function( x, y=NULL, method = "euclidean",
                         delta = 1,
                         upper = if(is.null(y)) FALSE else NULL,
                         p = 2, miles=TRUE, R=NULL
#                         eps =  NULL, diag = NULL
                         )
{
  # see help for exact parameter meaning

  # We always include all small distances. Hence, this function 
  #   works different than any other spam functions. An addititonal
  #   call to an as.spam would eliminate the small values. 
#  if (!is.null(diag)) warning("Argument "diag" is deprecated")
#  if (!is.null(eps))  warning("Argument "eps" is deprecated")
  
  if (!is.na(pmatch(method, "euclidian")))     method <- "euclidean"
  METHODS <- c("euclidean", "maximum", "minkowski", "greatcircle")
  method <- pmatch(method, METHODS)  # result is integer

  if (is.na(method))     stop("invalid distance method")

  force64 <- getOption("spam.force64")

  if (method == 4) {
    if (is.null(R))
      p <- ifelse( miles,3963.34,6378.388)
    else {
      if (R <= 0)           stop("'R' should be postiive")
      p <- R
    }
    if (abs(delta)>180.1)  stop("'delta' should be smaller than 180 degrees.")
  }
  

  if (is.null(upper)) 
    part <- 0L
  else
    part <- ifelse(upper, 1L ,-1L)
  if (is.data.frame(x))  x <- as.matrix(x)
  if (is.list(x)) stop("'x' should be an array or matrix")
           # as.matrix( list() ) does not work
  if (!is.matrix(x)) x <- as.matrix(x)
  nd <- dim(x)[2]
  n1 <- dim(x)[1]

  if (!is.null(y)) {
    # we specify x and y:
    if (is.data.frame(y))  y <- as.matrix(y)
    if (is.list(x)) stop("'x' should be an array or matrix")
    if (!is.matrix(y)) y <- as.matrix(y)
    if (nd!=dim(y)[2]) stop("'x' and 'y' should have the same number of columns.")
    n2 <- dim(y)[1]
    mi <- min(n1,n2)
    ma <- max(n1,n2)
    nnz <- min(max(getOption("spam.nearestdistnnz")[1],
                   ma*getOption("spam.nearestdistnnz")[2]),
               (as.double(mi)*(mi+1)+(ma-mi)^2)/ ifelse( is.null(upper), 1, 2))
    # there is an as.double just in case that mi (and n1 below) is > 2^16
  } else {
    # x = y, i.e. proper distance matrix
    if (n1==1)         stop("More than a single point in 'x' is required.")
    if (method == 4) {
      p <- -p  # we save one variable...
    }
    y <- x
    n2 <- n1
    nnz  <- min(max(getOption("spam.nearestdistnnz")[1],
                    n1*getOption("spam.nearestdistnnz")[2]),
                (as.double(n1)*(n1+1))/ ifelse( is.null(upper), 1, 2))
  }
  
  # EXPLICIT-STORAGE-FORMAT     
  if(max(n1,n2,nnz) > 2147483647 - 1 || force64)
      SS <- .format64()
  else
      SS <- .format32
  
  if(2147483647 < nnz) stop("Distance matrix is too dense (1)")

  repeat {
    d <- NULL # Free the memory allocated by a previous attemp
    d <- .C64("closestdist",
    ##              subroutine closestdist( ncol, x,nrowx, y, nrowy,
    ##  &    part, p, method, 
    ##  &    eta, colindices, rowpointers, entries, nnz, iflag)
              SIGNATURE=c(SS$signature, "double", SS$signature, "double", SS$signature,
                          SS$signature, "double", SS$signature, "double", SS$signature,
                          SS$signature, "double", SS$signature, SS$signature),
              
              nd,
              x,
              n1, #w
              y,
              
              n2, #w
              part, #arg 6
              p[1], 
              method, 
              
              abs( delta[1]),
              colindices=vector(SS$type, nnz),
              rowpointers=vector(SS$type, n1+1), #arg 11
              entries=vector("double",nnz), 
              
              nnz=nnz,
              iflag=0,
              
              INTENT = c("r", "r", "r", "r",
                     "r", "r", "rw", "r", 
                     "r", "w", "w", "w", 
                     "rw", "w"),
              NAOK = getOption("spam.NAOK"),
              PACKAGE=SS$package)
    
    if (d$iflag==0) break else {
      nnz <-  nnz*getOption("spam.nearestdistincreasefactor")*n1/(d$iflag-1)
        
      # EXPLICIT-STORAGE-FORMAT     
      if(max(n1,n2,nnz) > 2147483647 - 1 || force64)
        SS <- .format64()
      else
        SS <- .format32
        
      if(nnz > 2147483647) stop("Distance matrix is too dense (2)")
      
      madens <- d$iflag
      warning(paste("You ask for a 'dense' sparse distance matrix, I require one more iteration.",
                            "\nTo avoid the iteration, increase 'nearestdistnnz' option to something like\n",
                            "'options(spam.nearestdistnnz=c(",d$nnz,",400))'\n(constructed ",madens,
                            " lines out of ",n1,").\n",sep=""), call. = TRUE)
              
    }
  }
  
  length(d$entries) <- d$nnz
  length(d$colindices) <- d$nnz
  
  if(d$nnz == 0) {
    return(.newSpam(
      dimension=c(n1,n2),
      force64 = force64
    ))
  }
  
  return(.newSpam(
    entries=d$entries,
    colindices=d$colindices,
    rowpointers=d$rowpointers,
    dimension=c(n1,n2),
    force64 = force64
  ))
}

# in fields:
# rdist <- function (x1, x2) 

spam_rdist <- function(x1, x2, delta = 1) 
     nearest.dist(x1, y=x2,   delta = delta,  upper = NULL)

# in fields:
# rdist.earth <- function (x1, x2, miles = TRUE, R = NULL) 
spam_rdist.earth <- function(x1, x2, delta=1, miles = TRUE, R = NULL)
    nearest.dist( x1, y=x2, method = "greatcircle",
                         delta = delta, miles=miles, R=R,  upper = NULL)

